docs/zkapps/o1js/bitwise-gadgets.mdx

```diff
@@ -0,0 +1,320 @@
+- Interprets small "negative" field element inputs as large integers close to the field size, so they don't pass the 64-bit check. To prove that a value lies in the int64 range [-2^63, 2^63), use `rangeCheck64(x.add(1n << 63n))`.
+
+Example:
+
+```ts
+const x = Provable.witness(Field, () => Field(12345678n));
+Gadgets.rangeCheck64(x); // successfully proves 64-bit range
+
+const xLarge = Provable.witness(Field, () => Field(12345678901234567890123456789012345678n));
+Gadgets.rangeCheck64(xLarge); // throws an error since input exceeds 64 bits
+```
+
+## rightShift
+
+```ts
+rightShift(field: Field, bits: number) {
+    return rightShift(field, bits);
+}
+```
+
+The rightShift gadget is a provable method that supports the bitwise shifting operation that moves the bits of a binary number to the right. Unlike [rotate](#rotate), the bits that fall off at the end are discarded and the vacant positions are filled with zeros.
+
+The rightShift gadget:
+
+- Performs a right shift operation on a Field element.
+- Shifts bits to the right and discards the overflowing bits.
+- Performs these operations with the big-endian 64-bit representation of the number, where the most significant (64th) bit is on the left end and the least significant bit is on the right end.
+- Requires values that are range checked to 64 bits. You can use [rangeCheck64](#rangecheck64).
+- Throws an error if the input exceeds 64 bits.
+
+Example:
+```ts
+const x = Provable.witness(Field, () => Field(0b001100)); // 12 in binary
+const y = Gadgets.rightShift(x, 2); // right shift by 2 bits
+y.assertEquals(0b000011); // 3 in binary
+
+const xLarge = Provable.witness(Field, () => Field(12345678901234567890123456789012345678n));
+rightShift(xLarge, 32); // throws an error since input exceeds 64 bits
+```
+
+## rotate
+
+```ts
+rotate(field: Field, bits: number, direction: 'left' | 'right' = 'left') {
+return rotate(field, bits, direction);
+},
+```
+
+The rotate gadget operates similarly to the [Right shift (>>)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Right_shift) and [Left shift (<<)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Left_shift) operators in JavaScript. 
+
+The rotate gadget is a provable method to support bitwise left or right rotation for native Field elements. 
+
+The rotate gadget:
+
+- Shifts bits to the left or the right. 
+- The rotate left operation (ROL) shifts bits to the left. The bits that fall off the leftmost side wrap around and reappear on the rightmost side.
+- The rotate right operation (ROR) shifts bits to the right. The bits that fall off the rightmost side wrap around and reappear on the leftmost side.
+- Performs these operations with the big-endian 64-bit representation of the number, where the most significant (64th) bit is on the left end and the least significant bit is on the right end.
+- Requires values that are range checked to 64 bits. You can use [rangeCheck64](#rangecheck64).
```

Comment:
can use or must use? (asked question already for other gadgets, not sure if the answer if the same for all range checked values)